### PR TITLE
[2021.2][SG] Cleanup Blackboard Categories

### PR DESCRIPTION
**Pull request by Nightmask3**
---
### Purpose of this PR

This PR implements a number of fixes to the shader graph blackboard categories feature that was recently checked into master by this PR : #4452 

These are issues that were reported by an external partner as well as some issues that were raised by user testing after the previous PR landed.

The issues fixed by this PR:

image4

- Renaming a category would also select all children within that category, interfering with future operations such as adding another category

Fixed behavior:

https://user-images.githubusercontent.com/13033542/121451017-a011ff00-c951-11eb-81bb-e9a6f98fffa7.mp4

image1

- Was pointed out that its a bit hard to notice the drag indicator when its at the bottom of a category, this is made more prominent now

Fixed behavior: 

image

https://user-images.githubusercontent.com/13033542/121454678-19aceb80-c958-11eb-82d6-be8e1e8fd554.mp4

- An error introduced by the previous PR was an inability to use the Graph Settings menu in the Inspector when a category was selected due to it auto-switching to the node settings tab, this is fixed now

TQv4NhMbGm

- An issue when the ShaderGUI is called with a subset of properties – a thing that's common for custom shader GUIs, handle some properties that need specific handling and then call the ShaderGUI that would have been used originally with a filtered set of properties.
    - Integrated a PR submitted by an external partner: https://github.com/Unity-Technologies/Graphics/pull/4584 that fixes this by not throwing an exception when a property is not found, instead just continues iterating through the properties. 

- A ShaderGraph with no `ShaderGraphMetadata` asset breaks the Material inspector
    - This is fixed now, we simply just fall back on the previous material inspector behavior handling for shader properties if no `ShaderGraphMetadata` asset can be found

- Duplicating categories leaves selection in stuck state

Fixed behavior: 

https://user-images.githubusercontent.com/13033542/121452241-f1bb8900-c953-11eb-9c78-8ce8b08e9302.mp4

^^ Also fixed some other related issues with copy-pasting of properties/categories/etc where I realized the editor conventions of clearing current selection and then selecting the newly copied items post the duplication/paste operation were not being followed

- Moving properties around now has a slight insertion offset, doesn't match previous (expected) behaviour for where the insertion happens based on position above/below center of the hovered property

Fixed behavior: 

https://user-images.githubusercontent.com/13033542/121452334-20396400-c954-11eb-9f2b-4952ea161d11.mp4

^^ This issue also led me to discovering a host of other issues that existed with the handling of moving items within categories, where it was also not in line with the editor conventions set by the Scene Hierarchy list. We now follow those conventions exactly.

image

- Shader property order is messed up, not the same as in Blackboard

This was an issue introduced by the categories PR, and has been fixed now so that properties and keywords will remain in consistent order from the blackboard when getting inserted as property blocks into the shader. 

Bugs fixed by this PR:

- https://fogbugz.unity3d.com/f/cases/1337100/

---
### Testing status

- Testing all of the cases outlined above to ensure they no longer repro
- Extensive user testing across a variety of scenarios for inserting/moving items into and between categories:
    - Moving single items within a category, and save/load graph to ensure output is correct
    - Moving multiple consecutive items within a category, and save/load graph to ensure output is correct
    - Moving multiple non-consecutive items within a category, and save/load graph to ensure output is correct
    - Inserting single items from another category into a category, and save/load graph to ensure output is correct
    - Inserting multiple consecutive items from another category into a category, and save/load graph to ensure output is correct
    - Inserting multiple non-consecutive from another category into a category, and save/load graph to ensure output is correct
    - Combining the various above permutations for selections of items within the same category and items outside the category, to ensure that the order of items on the blackboard is persisted even after the move/insert operation
        - What this means is that, we are now sorting the items that are being moved/inserted such that 
             - the items are inserted into their destination category in such a way that the items that are within the category will always be inserted first
             - then items from other categories will be inserted after
                 - For items from other categories, their relative order across the blackboard will be preserved as they get inserted into the destination category
                 - This gif below should demonstrate the idea, notice that the items being inserted from the other categories, are retaining the order of the category that they come from, relative to the destination category, prior to this the items were inserted in the order that they were selected, which was not robust and inconsistent feeling as far as the UX goes. 

https://user-images.githubusercontent.com/13033542/121453981-e6b62800-c956-11eb-8d7c-0ce7c7de9fc1.mp4

---
### Comments to reviewers

Doesn't include a changelog because none of these are changes to anything that was user facing since the last release, just fixing stuff intended for the next release (i.e. 21.2).

### CODE REVIEW

**Review by GrantLamb-Unity — APPROVED** [review]
Thank you for the before and after videos for these bugfixes!

File: com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs [hunk 1]
```diff
@@ -26,7 +26,7 @@ namespace UnityEditor.ShaderGraph
     // This ifdef can be removed once V2 is the only option.
     [ScriptedImporter(115, Extension, -902)]
 #else
-    [ScriptedImporter(48, Extension, -902)]
+    [ScriptedImporter(49, Extension, -902)]
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs:29]: Why only update one of these numbers? You need to update both if the serialization actually requires a reimport on upgrade.

**Review by marctem — CHANGES_REQUESTED** [review]
One potential issue with the importer and then I think it's good to land!

File: com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs (hunk 1 above)
Comment (elizabeth-legros) [reply com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs:29]: correct, both should be updated. My mistake :)

Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs:29]: Thanks Liz! Will amend this 👍

**Comment by Nightmask3** [comment]
Test run with most jobs passing, one test failure and formatting failure: 
image

Ran the formatter (pass of the formatter test after my last commit can be seen here on Github)

And also fixed the sole test failure and reran that job, which succeeded:
image

File: com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs (hunk 1 above)
Comment (marctem) [reply com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs]: Ah, no, you changed the entirely wrong thing. you need:

```
 [ScriptedImporter(116, Extension, -902)]
#else
    [ScriptedImporter(49, Extension, -902)]
```

The first number is the call is the version, unique to SG, which must be bumped by 1 independently for each branch of the #if. The second number is the global importer order which should be the same for both branches of the #if, and should not change.

Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs]: Ahh that makes sense 👍 
Currently in master the state of the Importer parameters is as such:

```
    [ScriptedImporter(121, Extension, -902)]
#else
    [ScriptedImporter(50, Extension, -902)]
#endif
```

post the merge into this branch and bumping the appropriate version numbers: 

```
    [ScriptedImporter(122, Extension, -902)]
#else
    [ScriptedImporter(51, Extension, -902)]
```

**Review by marctem — APPROVED** [review]

**Pull request merged**